File: src/main/java/Rover.java
public class Rover {

    public static final int MAX_WIDTH = 10;
    public static final int MAX_HEIGHT = 10;

    private Coordinate direction = Coordinate.NORTH;

    private int x;

    private int y;

    public String execute(String commands) {
        commands.chars()
                .mapToObj(i -> (char) i)
                .forEach(this::calculatePosition);
        return x + ":" + y +  ":" + this.direction.value();
    }

    private void calculatePosition(char command) {
        if (command == 'R') {
            turnRight();
        } else if (command == 'L') {
            turnLeft();
        } else if (command == 'F') {
            moveForward();
        }
    }

    private void moveForward() {
        if (this.direction == Coordinate.EAST) {
            this.x = (x + 1) % MAX_WIDTH;
        } else if (this.direction == Coordinate.WEST) {
            this.x = this.x == 0 ? 9 : this.x - 1;
        } else if (this.direction == Coordinate.NORTH) {
            this.y = this.y == 0 ? 9 : this.y - 1;
        } else if (this.direction == Coordinate.SOUTH) {
            this.y = (y + 1) % MAX_HEIGHT;
        }
    }

    private void turnLeft() {
        this.direction = direction.left();
    }

    private void turnRight() {
        this.direction = direction.right();
    }

}

File: scala/src/main/scala/Rover.scala

class Rover(grid: Grid) {

  var direction: Direction = new North();

  var position = Position(0, 0)

  var obstacleFound = false

  def execute(commands: String): String =
    commands
      .toCharArray
      .map(command => processCommand(command))
      .last


  def processCommand(command: Char): String = {
    if (!obstacleFound) {
      if (command == 'R') turnRight()
      if (command == 'L') turnLeft()
      if (command == 'F') moveForward(command)
      if (command == 'B') moveBackward(command)
    }
    formatResult()
  }

  def turnRight(): Unit = {
    direction = grid.right(direction)
  }

  def turnLeft(): Unit = {
    direction = grid.left(direction)
  }

  def moveForward(command: Char): Unit = {
    val newPosition: Position = direction.forward(position)
    checkObstacles(newPosition)
  }

  def moveBackward(command: Char): Unit = {
    val newPosition: Position = direction.backward(position)
    checkObstacles(newPosition)
  }

  def checkObstacles(newPosition: Position): Unit = {
    if (grid.obstacleIn(newPosition)) obstacleFound = true
    else position = newPosition
  }

  def formatResult(): String = {
    val prefix = if (obstacleFound) "O:" else ""
    s"$prefix${position.x}:${position.y}:${direction.value}"
  }

}
